apps/agentic-chat/src/hooks/useTools.ts

```diff
@@ -91,368 +36,72 @@ const useTools = () => {
-          });
-
-          return hash;
-        } catch (err) {
-          console.error('Error approving token', err);
-          throw err;
-        }
+        return approve(walletClient, toolCall);
       }
 
       case 'sendTransaction': {
         const typedToolCall = toolCall as ToolCall<
           'sendTransaction',
           {
             to: Address;
-            value: string;
+            valueCryptoPrecision: string;
             data: Hex;
             chainId: number;
           }
         >;
-        const { to, value, data, chainId } = typedToolCall.args;
 
-        const publicClient = getPublicClient(chainId);
-        const account = walletClient?.account;
-        if (!walletClient || !account?.address) {
-          throw new Error('No account connected');
-        }
+        const { to, valueCryptoPrecision, data, chainId } = typedToolCall.args;
 
-        try {
-          // First estimate gas to catch potential errors
-          const [gasLimit, gasPrice] = await Promise.all([
-            publicClient.estimateGas({
-              account,
-              to: getAddress(to),
-              value: BigInt(value),
-              data: data,
-            }),
-            publicClient.getGasPrice(),
-          ]);
+        const valueCryptoBaseUnit = toBaseUnit(
+          valueCryptoPrecision,
+          18 // Assuming 18 decimals for ETH-like transactions
+        );
 
-          // Now send the transaction with the estimated gas
-          const hash = await walletClient.sendTransaction({
-            account,
-            to: getAddress(to),
-            value: BigInt(value),
-            data: data,
-            chain: getChainById(chainId),
-            gas: gasLimit,
-            gasPrice,
-          });
-          return hash;
-        } catch (err) {
-          console.error('Error sending transaction', err);
-          throw err;
-        }
+        return sendTransaction({
+          walletClient,
+          value: valueCryptoBaseUnit,
+          data,
+          chainId,
+          to,
+        });
       }
 
       case 'executeSwap': {
         if (!bebopQuote) {
           throw new Error('No quote available');
         }
 
-        const { tx } = bebopQuote;
-
-        const { to, value, data, chainId } = tx;
-
-        const publicClient = getPublicClient(chainId);
-        const account = walletClient?.account;
-        if (!walletClient || !account?.address) {
-          throw new Error('No account connected');
-        }
-
-        try {
-          // First estimate gas to catch potential errors
-          const [gasLimit, gasPrice] = await Promise.all([
-            publicClient.estimateGas({
-              account,
-              to: getAddress(to),
-              value: BigInt(value),
-              data: data,
-            }),
-            publicClient.getGasPrice(),
-          ]);
+        const { chainId, tx } = bebopQuote;
+        const { to, value, data } = tx;
```

Comment:
Logic not fully extracted so we can keep `sendTransaction` speaking raw `value`, but have this as a higher-order function doing the quote to sendTransaction wrapping, keeping core logic in one place.